cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/events/KubernetesRestartEventPublisher.java

```diff
@@ -0,0 +1,139 @@
+/*
+ * Copyright Strimzi authors.
+ * License: Apache License 2.0 (see the file LICENSE or http://apache.org/licenses/LICENSE-2.0.html).
+ */
+package io.strimzi.operator.cluster.operator.resource.events;
+
+import io.fabric8.kubernetes.api.model.MicroTime;
+import io.fabric8.kubernetes.api.model.ObjectReference;
+import io.fabric8.kubernetes.api.model.ObjectReferenceBuilder;
+import io.fabric8.kubernetes.api.model.Pod;
+import io.fabric8.kubernetes.client.KubernetesClient;
+import io.strimzi.operator.cluster.model.RestartReason;
+import io.strimzi.operator.cluster.model.RestartReasons;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+import java.time.Clock;
+import java.time.ZonedDateTime;
+import java.util.Locale;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import static java.nio.charset.StandardCharsets.UTF_8;
+
+/**
+ * Publishes KafkaRoller restart events as Kubernetes events, using either the events.k8s.io/v1beta1 or v1 API,
+ * depending on which is available.
+ */
+public abstract class KubernetesRestartEventPublisher {
```

Comment:
What is the dependency on pod restarts? Should this be designed as `KubernetesEventPublisher` independently on the restart events? If nothing else, the name here conflicts with the generic name of the implementations, so one or the other is wrong. I would probably say to make it as generic as possible.

Reply:
At the moment, it requires a `RestartReasons` to be provided in the `publishRestartEvents` method, so it's rather targeted right now, but I'm happy to rename it to `KubernetesEventPublisher`, as it could be easily extended to publish, well, any other event we want to.